### ISSUE TITLE
Feature: Add hooks for custom races [$15]

### ISSUE THREAD

**Issue by heyitsbench** [azerothcore/azerothcore-wotlk#7648]
### Describe your feature request or suggestion in detail

Add hooks in SharedDefines needed to add custom races similarly to this commit:
https://github.com/benjymansy123/azerothcore-wotlk/commit/3f161cf01266267077b721e2dbdcb01696cfc903

[Bountysource](https://app.bountysource.com/issues/100762579-feature-add-hooks-for-custom-races)

**Comment by lineagedr** [comment]
😄

**Comment by T1ti** [comment]
no it's a bad idea to enable custom races for a generic blizzlike emulator, just enable them by yourself on your local

**Comment by benjymansy123** [comment]
I'm sorry, but couldn't the case be made that _many_ of the modules currently featured on the catalogue on the AzerothCore website are very much not blizzlike?

**Comment by T1ti** [comment]
that's exactly why they are modules
Enabling the custom races by default could create securitiy issues, some users could be able to create custom races by sending custom packets

**Comment by benjymansy123** [comment]
The point of this issue is to be able to make custom races into a module, not enable them by default. Currently, you can't make custom races into a module, you'd have to change the SharedDefines file yourself, which I'm sure many would agree is far from convenient.

**Comment by lineagedr** [comment]
It should be possible to make this work without adding new races. Perhaps if we convert these macros into a uint32

RACEMASK_ALL_PLAYABLE
RACEMASK_ALLIANCE

And then in the module we define a new enum for the new races and add the new flags to the RACEMASK_ALL_PLAYABLE

**Comment by T1ti** [comment]
If you're enabling custom races, having to set the race mask yourself is 0.01% of the work.

**Comment by benjymansy123** [comment]
Yes, it is 0.01% of the work, but I'm in the process of doing the rest of it. Turning this into a module has seen interest from a good several people in the discord, but I don't know how else to do it without a hook.

**Comment by lineagedr** [comment]
SharedDefines.h
We keep the same blizzlike races we just convert the macros to a datatype that can be modified at runtime.
```c++
uint8 MAX_RACES = 12;

uint16 RACEMASK_ALL_PLAYABLE = ((1 <<  (RACE_HUMAN    - 1)) | (1 << (RACE_ORC           - 1)) | (1 << (RACE_DWARF    - 1)) |
                                (1 <<  (RACE_NIGHTELF - 1)) | (1 << (RACE_UNDEAD_PLAYER - 1)) | (1 << (RACE_TAUREN   - 1)) |
                                (1 <<  (RACE_GNOME    - 1)) | (1 << (RACE_TROLL         - 1)) | (1 << (RACE_BLOODELF - 1)) |
                                (1 <<  (RACE_DRAENEI  - 1)));

uint16 RACEMASK_ALLIANCE = ((1 << (RACE_HUMAN - 1)) | (1 << (RACE_DWARF   - 1)) | (1 << (RACE_NIGHTELF - 1)) | 
                           (1 << (RACE_GNOME  - 1)) | (1 << (RACE_DRAENEI - 1)));

uint16 RACEMASK_HORDE = RACEMASK_ALL_PLAYABLE & ~RACEMASK_ALLIANCE;
```

Your module file.
```c++
#include "SharedDefines.h"

// New races
enum class NewRaces : uint8
{
    RACE_GOBLIN = 9,
    RACE_WORGEN = 12,
    RACE_MAX    = 13
}

class MyRaceModule : public WorldScript 
{
public:
    MyRaceModule() 
        : WorldScript("MyRaceModule") { }

    // Hopefully this hook works and we don't need to add a new one that executes on program launch
    void OnStartup() override
    {
        // Let's add goblin and worgen to the playable racemask
        RACEMASK_ALL_PLAYABLE |= (1 << (NewRaces::RACE_GOBLIN - 1)) 
                              |  (1 << (NewRaces::RACE_WORGEN - 1));

        // Add them to their respectable faction
        RACEMASK_HORDE        |= (1 << (NewRaces::RACE_GOBLIN - 1));
        RACEMASK_ALLIANCE     |= (1 << (NewRaces::RACE_WORGEN - 1));

        // 2 new races
        MAX_RACES += (NewRaces::RACE_MAX - 1);
    }
};

```

**Comment by benjymansy123** [comment]
May I ask why this was closed?

**Comment by yuanf225** [comment]
Is there any progress now for custom races? Is anyone interested in this

**Comment by benjymansy123** [comment]
https://github.com/benjymansy123/custom-race-ac-12_6_21

**Comment by yuanf225** [comment]
> https://github.com/benjymansy123/custom-race-ac-12_6_21

https://model-changing.net/files/file/191-aplharetail-worgen-and-goblin-335/
You try this

**Comment by TheLegendaryKingdom** [comment]
I think the ambition behind this request is the availability of addingNEW races to the game, not replacing existing ones (:

**Comment by benjymansy123** [comment]
If anyone is interested, it looks like something similar to what was done [here](https://github.com/azerothcore/azerothcore-wotlk/pull/1444) for battlegrounds could be done for races as well.

**Comment by Helias** [comment]
> SharedDefines.h We keep the same blizzlike races we just convert the macros to a datatype that can be modified at runtime.
> 
> ```c++
> uint8 MAX_RACES = 12;
> 
> uint16 RACEMASK_ALL_PLAYABLE = ((1 <<  (RACE_HUMAN    - 1)) | (1 << (RACE_ORC           - 1)) | (1 << (RACE_DWARF    - 1)) |
>                                 (1 <<  (RACE_NIGHTELF - 1)) | (1 << (RACE_UNDEAD_PLAYER - 1)) | (1 << (RACE_TAUREN   - 1)) |
>                                 (1 <<  (RACE_GNOME    - 1)) | (1 << (RACE_TROLL         - 1)) | (1 << (RACE_BLOODELF - 1)) |
>                                 (1 <<  (RACE_DRAENEI  - 1)));
> 
> uint16 RACEMASK_ALLIANCE = ((1 << (RACE_HUMAN - 1)) | (1 << (RACE_DWARF   - 1)) | (1 << (RACE_NIGHTELF - 1)) | 
>                            (1 << (RACE_GNOME  - 1)) | (1 << (RACE_DRAENEI - 1)));
> 
> uint16 RACEMASK_HORDE = RACEMASK_ALL_PLAYABLE & ~RACEMASK_ALLIANCE;
> ```
> 
> Your module file. EXAMPLE CODE
> 
> ```c++
> #include "SharedDefines.h"
> 
> // New races
> enum class NewRaces : uint8
> {
>     RACE_GOBLIN = 9,
>     RACE_WORGEN = 12
> }
> 
> class MyRaceModule : public WorldScript 
> {
> public:
>     MyRaceModule() 
>         : WorldScript("MyRaceModule") { }
> 
>     // Hopefully this hook works and we don't need to add a new one that executes on program launch
>     void OnStartup() override
>     {
>         // Let's add goblin and worgen to the playable racemask
>         RACEMASK_ALL_PLAYABLE |= (1 << (NewRaces::RACE_GOBLIN - 1)) 
>                               |  (1 << (NewRaces::RACE_WORGEN - 1));
> 
>         // Add them to their respectable faction
>         RACEMASK_HORDE        |= (1 << (NewRaces::RACE_GOBLIN - 1));
>         RACEMASK_ALLIANCE     |= (1 << (NewRaces::RACE_WORGEN - 1));
> 
>         // Add 2 new races
>         MAX_RACES += 2;
>     }
> };
> ```

this didn't not work for me, we need another approach

**Comment by heyitsbench** [comment]
Linked PR does not add hooks, but it does close this issue (as it dynamically calculates the `SharedDefines` pieces from DBC data instead of relying on hardcoded values).
